### PR TITLE
fix(cursor): native cursor overlay fixes for multi-display, crop, zoom and export consistency

### PR DESCRIPTION
**Pull request by auberginewly**
## Summary

A series of fixes for the macOS native cursor overlay across the
multi-display recording, video crop, zoom, rounded-corner and export
scenarios — keeping the **preview path (VideoPlayback.tsx)** and the
**export path (frameRenderer.ts)** consistent — plus a user-facing
toggle to allow the cursor to overflow the canvas.

### Multi-display ghost cursor & motion trail
Recording display A, moving the mouse to non-recorded display B clamped
the cursor to the canvas edge and kept it visible, producing a long
trail flying back from the edge on return.
- In `macNativeCursorRecordingSession`, mark `visible: false` when out of
  bounds — this triggers `hideNativeCursorPreview()` and resets smoothing /
  motion blur, so the cursor snaps cleanly on return.
- Add hysteresis (only hide after 3 consecutive out-of-bounds samples,
  ≈100ms); fast swipes out-and-back are handled by clip-path at the edge
  instead of flickering invisible.

### Cursor overflowing the video canvas
The enlarged cursor overlay bled into the editor background. The
ancestor's `overflow:hidden` was broken by `transform-style: preserve-3d`
on `composite3DRef`.
- Switch to dynamic `clip-path: inset()` (unaffected by preserve-3d),
  clipping precisely via `baseMaskRef` (incl. crop) + border radius.
- Move `nativeCursorClipRef` out of the `preserve-3d` container so
  clip-path stays correct under 3D-rotation zoom regions.
- Make the clip-path camera-aware so it follows cameraContainer
  scale/translation.

### Preview / export cursor size mismatch
`renderAsset.width` is absolute px; preview canvas ~600px vs export
1920px gave a ~3x size discrepancy.
- Both paths now normalize by `maskRect.width / croppedVideoWidth`, so
  the cursor occupies the same fraction of the video in both.

### Lost rounded corners & zoom over-crop on export
- Export goes through Canvas 2D; PIXI WebGL mask alpha is unreliable at
  corners → explicit `roundRect` clip enforces rounded corners (shadow
  path erases square corners via destination-out + evenodd, then
  redraws).
- **Export cropped off part of the video while zoomed**: the PIXI mask
  is inside cameraContainer and scales with zoom so the video fills the
  magnified bounds, but the static maskRect rounded clip on export still
  used the small non-zoom size. Extracted `cameraAwareMaskRect()` to
  share one camera transform across shadow / direct / cursor paths;
  degrades to the original static behavior at zoom=1 (no regression).

### Consistency cleanup
- Cache `maskBorderRadius` in `LayoutCache` so preview and export share
  the same rounded-corner value.
- Remove a conflicting `borderRadiusRef` sync effect: a standalone
  effect overwrote the screenCover→0 (full-bleed) corrected value with
  the raw prop, making vertical-stack/split presets disagree between
  preview and export.

## New feature: "Clip to Canvas" toggle

Clipping the cursor to the video bounds is now optional. Added a
`cursorClipToBounds` setting (default **on**, fully backward compatible)
in the Show Cursor settings section:

- Default on → cursor is clipped to the video canvas (current behavior).
- Off → the cursor is allowed to overflow into the background, in both
  the preview and the exported MP4/GIF.
- Gating is applied **only** to the cursor clip (preview `clip-path`,
  export `drawNativeCursor`); the video rounded-corner clips in
  `compositeWithShadows` are fully decoupled and unaffected.
- Session-only, not persisted (consistent with the other cursor
  settings).

The entire cursor settings section (Show Cursor / Size / Smoothing /
Motion Blur / Click Bounce / Clip to Canvas) is now i18n-backed across
all 11 locales (previously hardcoded English).

## Key files

| File | Change |
|------|--------|
| `electron/native-bridge/cursor/recording/macNativeCursorRecordingSession.ts` | out-of-bounds `visible:false` + hysteresis |
| `src/components/video-editor/VideoPlayback.tsx` | dynamic camera-aware clip-path, moved out of preserve-3d, size normalization, removed redundant effect, clip gated by toggle |
| `src/components/video-editor/videoPlayback/layoutUtils.ts` | expose maskBorderRadius on LayoutResult |
| `src/lib/exporter/frameRenderer.ts` | cameraAwareMaskRect() helper, rounded corners, camera-aware clip, size normalization, cursor clip gated by toggle |
| `src/components/video-editor/{VideoEditor,SettingsPanel}.tsx`, `types.ts` | cursorClipToBounds state + Switch UI + i18n |
| `src/lib/exporter/{videoExporter,gifExporter}.ts` | thread cursorClipToBounds into export config |
| `src/i18n/locales/*/settings.json` (11 locales) | new `cursor` section |

## Test plan

- [x] Single display: cursor near the edge is clipped, no bleed into editor background; clipped to rounded corners with border radius; confined to cropped area with crop
- [x] Multi-display: moving to non-recorded display hides the cursor immediately; returning shows it with no trail; fast swipe out-and-back does not flicker invisible
- [x] Zoom: cursor stays inside the video bounds in both preview and exported mp4; zoom=1 behaves as before
- [x] Rounded corners + zoom: export keeps correct rounded corners and does not over-crop content
- [x] Preview vs export: cursor occupies the same fraction of the video
- [x] "Clip to Canvas" toggle off: cursor overflows the canvas in both preview and exported MP4/GIF; video rounded corners still correct
- [x] "Clip to Canvas" toggle on (default): behavior identical to before
- [x] Cursor settings labels are translated when switching language (11 locales)
- [x] `npm run i18n:check`, `tsc --noEmit`, `npm run build-vite` all pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added cursor clipping option in settings to control whether the cursor is confined to video canvas bounds

* **Improvements**
  * Enhanced cursor visibility tracking when cursor moves outside display boundaries
  * Improved cursor rendering consistency between on-screen preview and exported videos

[Review Change Stack](https://app.coderabbit.ai/change-stack/EtienneLescot/openscreen/pull/2?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by auberginewly** [comment]
现在我是已经把光标 overflow 给删掉了 修了多余运动轨迹采集的 bug 我来加一个可选 overflow or not 的开关

After more testing I've reconsidered the cursor-overflow approach:

Removed the overflow clipping. The hard clip-path/canvas clipping that confined the cursor to the video bounds is gone — clipping the native cursor at the edge looked worse than letting it sit slightly outside in some cases.
Fixed the real root cause: redundant motion-trail sampling. The phantom trail wasn't really an overflow problem — it came from stale out-of-bounds samples being collected and fed into smoothing. That's now fixed at the sampling layer, so the trail/ghosting is gone without needing to clip.
Next: make overflow opt-in. I'll add an optional "clip cursor to canvas / allow overflow" toggle so the behavior is configurable instead of hard-coded, rather than forcing one choice on everyone.
Will push the toggle commit shortly.

**Comment by auberginewly** [comment]
@EtienneLescot

This part is done 👇

Multi-display ghost cursor / motion-trail fix
Cursor-overflow clipping (preserve-3d / camera-aware / rounded corners / crop)
Preview vs export cursor size normalization
Lost rounded corners & zoom over-crop on export — fixed
New optional "Clip to Canvas" toggle (default on, backward compatible; turn off to let the cursor overflow the canvas)
Entire cursor settings section now i18n-backed across 11 locales
i18n:check / tsc / build-vite all pass locally.

Flow: please @EtienneLescot review and merge this PR into feat/macos-native-capture-pipeline first; it then flows upstream into main via #573. Ready for review — thanks! 🙏

**Comment by EtienneLescot** [comment]
@auberginewly  Thanks for the contribution, the feature looks relevant and useful to me overall.

I’d like to keep this PR focused and easy to review before merging it into the macOS native capture branch. Could you please squash/reorganize the current commits into 1 or 2 clean commits? There are currently quite a few intermediate commits for the size of the change, and a cleaner history would make the contribution easier to understand.

One technical point to double-check before merge: in `cameraAwareMaskRect()` inside `src/lib/exporter/frameRenderer.ts`, the export path clamps the transformed mask bounds to the stage using `Math.max(0, ...)` / `Math.min(stageW/stageH, ...)`. In the preview path, the CSS `clip-path` can naturally use bounds that extend outside the canvas. Could you verify whether this can create a preview/export mismatch when zoom/pan pushes the mask outside the stage, especially with rounded corners? If needed, please adjust the export behavior so it matches preview.

Apart from that, the functional scope looks good to me: cursor canvas clipping, editor/export consistency, and the related cursor settings i18n all seem relevant.

**Comment by auberginewly** [comment]
Done — two things addressed:

**Commits squashed**
The 20 intermediate commits are now 2 clean ones on `fix/cursor-canvas-clip`:
- `feat(cursor): clip native cursor to camera-aware video bounds in preview and export`
- `feat(cursor): add cursorClipToBounds setting with i18n translations`

**`cameraAwareMaskRect()` inconsistency — confirmed and fixed**
The concern was valid. The export path was clamping the transformed mask bounds to stage dimensions via `Math.max(0, …)` / `Math.min(stageW/stageH, …)`, while the preview's CSS `inset()` clip-path naturally accepts out-of-bounds values.

The specific mismatch with rounded corners: when zoom/pan pushes the mask rect partially off-stage (e.g. left edge is at `x = -30`), the clamped export path would start the `roundRect` clip at `x = 0` — producing a visible rounded corner at the stage edge. The preview has no such rounding there (the browser straight-cuts at the stage edge). This removed the `stageW/stageH` clamping and lets Canvas 2D clip naturally to its bounds, mirroring CSS behaviour:

```ts
// before
const left  = Math.max(0,      camX + camS * maskX);
const right = Math.min(stageW, camX + camS * (maskX + maskW));

// after — canvas clips out-of-bounds naturally, same as CSS inset()
x:     camX + camS * maskX,
width: camS * maskW,
```

The fix applies to all three call-sites (cursor clip, shadow composite, direct composite).

**Comment by EtienneLescot** [comment]
@auberginewly thanks for your swift reply ;-)
I let you see code rabbit comments and we should be all good!

**Comment by auberginewly** [comment]
> @auberginewly thanks for your swift reply ;-) I let you see code rabbit comments and we should be all good!

I see！Thanks！this is my first time to write a new feature，and I’ll try my best☺️

**Comment by auberginewly** [comment]
is that all right？👀

File: src/components/video-editor/SettingsPanel.tsx [hunk 1]
```diff
@@ -1381,86 +1385,102 @@ export function SettingsPanel({
...
-																onValueChange={(values) => onCursorSizeChange?.(values[0])}
-																min={0.5}
-																max={10}
-																step={0.1}
-																className="w-full [&_[role=slider]]:bg-[#34B27B] [&_[role=slider]]:border-[#34B27B] [&_[role=slider]]:h-3 [&_[role=slider]]:w-3"
+															<Switch
+																checked={cursorClipToBounds}
+																onCheckedChange={onCursorClipToBoundsChange}
+																className="data-[state=checked]:bg-[#34B27B] scale-90"
 															/>
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/components/video-editor/SettingsPanel.tsx:1408]: Addressed in `7cf8791`: the `cursorClipToBounds` switch now has an accessible label via `aria-label={t("cursor.clipToBounds")}`.

File: src/components/video-editor/VideoPlayback.tsx [hunk 2]
```diff
@@ -2006,6 +2033,27 @@ const VideoPlayback = forwardRef<VideoPlaybackRef, VideoPlaybackProps>(
...
+						ref={nativeCursorImageRef}
+						alt=""
+						aria-hidden="true"
+						className="absolute left-0 top-0 select-none"
+						style={{
+							display: "none",
+							pointerEvents: "none",
+							transformOrigin: "0 0",
+						}}
+					/>
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/components/video-editor/VideoPlayback.tsx:2061]: Addressed in `7cf8791`: `nativeCursorClipRef` now mirrors the `composite3DRef` transform during 3D rotation regions, while staying outside `preserve-3d` so the camera-aware `clip-path` continues to work.

File: src/i18n/locales/vi/settings.json [hunk 3]
```diff
@@ -191,6 +191,14 @@
...
 		"errorLoadFailed": "Không thể tải phông chữ. Vui lòng xác minh URL Google Fonts là chính xác."
 	},
+	"cursor": {
+		"show": "Hiện con trỏ",
+		"size": "Kích thước",
+		"smoothing": "Làm mượt",
+		"motionBlur": "Mờ chuyển động",
+		"clickBounce": "Nảy khi nhấp",
+		"clipToBounds": "Cắt theo khung"
+	},
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/i18n/locales/vi/settings.json:201]: Addressed in `7cf8791`: Vietnamese `cursor.motionBlur` now matches the existing `effects.motionBlur` wording (`Làm mờ chuyển động`).

File: src/i18n/locales/zh-TW/settings.json [hunk 4]
```diff
@@ -192,6 +192,14 @@
...
 		"errorLoadFailed": "無法載入該字體。請確認 Google Fonts URL 是否正確。"
 	},
+	"cursor": {
+		"show": "顯示游標",
+		"size": "大小",
+		"smoothing": "平滑",
+		"motionBlur": "運動模糊",
+		"clickBounce": "點擊彈跳",
+		"clipToBounds": "裁切至畫布"
+	},
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/i18n/locales/zh-TW/settings.json:202]: Addressed in `7cf8791`: Traditional Chinese `cursor.motionBlur` now matches the existing `effects.motionBlur` wording (`動態模糊`).

File: src/lib/exporter/frameRenderer.ts [hunk 5]
```diff
@@ -520,6 +522,30 @@ export class FrameRenderer {
...
+		const top = Math.max(0, camY + camS * maskY);
+		const right = Math.min(stageW, camX + camS * (maskX + maskW));
+		const bottom = Math.min(stageH, camY + camS * (maskY + maskH));
+		return {
+			x: left,
+			y: top,
+			width: right - left,
+			height: bottom - top,
+			br: this.layoutCache.maskBorderRadius * camS,
+		};
```
[the remark this answers is not in the document]
Comment (EtienneLescot) [reply src/lib/exporter/frameRenderer.ts:544]: Addressed in `c7d82db`: `cameraAwareMaskRect()` no longer clamps the transformed mask to the stage; Canvas now clips out-of-bounds naturally, matching the preview `inset()` behavior.

**Comment by EtienneLescot** [comment]
@auberginewly yes, this looks good to me now. I checked the latest commit `7cf8791`: the accessibility/i18n comments are addressed, and the 3D preview/export consistency fix is minimal and aligned with the existing transform flow. I also replied to the CodeRabbit threads with the relevant commit references. Thanks for the quick iteration!
